homeassistant/components/climate/oem.py

```diff
@@ -0,0 +1,149 @@
+"""
+OpenEnergyMonitor Thermostat Support.
+
+This provides a climate component for the ESP8266 based thermostat sold by
+OpenEnergyMonitor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/climate.openenergymonitor/
+"""
+import logging
+
+import voluptuous as vol
+
+# Import the device class from the component that you want to support
+from homeassistant.components.climate import (
+    ClimateDevice, PLATFORM_SCHEMA, STATE_HEAT, STATE_IDLE, ATTR_TEMPERATURE)
+from homeassistant.const import (CONF_HOST, CONF_USERNAME, CONF_PASSWORD,
+                                 CONF_PORT, TEMP_CELSIUS, CONF_NAME)
+import homeassistant.helpers.config_validation as cv
+
+# Home Assistant depends on 3rd party packages for API specific code.
+REQUIREMENTS = ['oemthermostat==1.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+# Local configs
+CONF_TARGET_TEMP = 'target_temp'
+CONF_AWAY_TEMP = 'away_temp'
+
+# Validation of the user's configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(
+        CONF_PORT, default=80): cv.port,
+    vol.Optional(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_TARGET_TEMP): vol.Coerce(float),
+    vol.Optional(CONF_AWAY_TEMP, default=14): vol.Coerce(float)
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup oemthermostat."""
+    from oemthermostat import Thermostat
+
+    # Assign configuration variables. The configuration check takes care they
+    # are present.
+    name = config.get(CONF_NAME)
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    target_temp = config.get(CONF_TARGET_TEMP)
+    away_temp = config.get(CONF_AWAY_TEMP)
+
+    therm = Thermostat(host, port=port,
+                       username=username, password=password)
```

Comment:
Usually, we'll want to catch exceptions from the underlying lib, and just set the relevant states properties to `None`. This means that if the device is disconnected when running, we won't be filling the log with exceptions. If it can't connect when setting up the platform, we usually fail the setup.

One Example below. If you dig through the source you'll find others:

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor
/darksky.py#L137